ADBDEV-3914 Fix plpgsql/sql subqueries logging

dnskvlnk opened the pull request:
gpdb changes gp_enable_gpperfmon guc value depending on log level
to control which statements to log during query execution. We don't
want to log nested statements of plpgsql function if log level is
other than DEBUG5, but gpperfmon logs some despite of log level.
This happened because only executor queries were properly covered
to have correct gp_enable_gpperfmon value, when utility queries 
were executed with wrong value of this guc.

This patch covers utility queries as well, setting gp_enable_gpperfmon
to correct value for them. Though gpperfmon doesn't log utility
queries, gp_enable_gpperfmon can be set for nested queries.

Also, gp_enable_gpperfmon was immediately reset in fmgr_sql to its
original value which lead incorrect logging of child queries. This
patch removes extra assignment.

Commit:
Remove extra gp_enable_gpperfmon assignment

Stolb27 commented on src/backend/executor/functions.c:
```diff
@@ -1278,8 +1278,6 @@ PG_TRY();
 		gp_enable_gpperfmon = false;
 	}
 
-	gp_enable_gpperfmon = orig_gp_enable_gpperfmon;
```
Good catch for conflict resolution error from upstream. BUT! `fmgr_sql` function is about `language sql` functions. Did you test your solution against original reproduction steps with `language plpgsql`? Can you explain how plpgsql function call gets here?

dnskvlnk replied:
Yes, `fmgr_sql` handles only regular sql functions. `Plpgsql` and `do` statements are handled by `plpgsql_exec_function`, I saved `gp_enable_gpperfmon` at the beginning of that function and restored it at the end which enabled correct logging for such functions and its contents

Stolb27 reviewed (changes requested):
Provide valid commit description first.

Commit:
Fix logging for plpgsql functions and 'do' statements

This patch fixes gpperfmon logging for both sql and pgplsql functions.
Before this fix, contents of such functions were logged incorrectly
leaving records in queries_history dated to 1970 year. This happened
because gp_enable_gpperfmon was set incorrectly to true for such
functions. This patch sets this variable to correct value to prevent
such behavior.

Stolb27 commented on src/pl/plpgsql/src/pl_exec.c:
```diff
@@ -492,6 +496,8 @@ plpgsql_exec_function(PLpgSQL_function *func, FunctionCallInfo fcinfo,
 	plpgsql_destroy_econtext(&estate);
 	exec_eval_cleanup(&estate);
 
+	gp_enable_gpperfmon = orig_gp_enable_gpperfmon;
```
What about `elog(ERROR` inside a plpgsql function?

dnskvlnk replied:
Fixed

Stolb27 commented on src/pl/plpgsql/src/pl_exec.c:
```diff
@@ -260,6 +260,10 @@ plpgsql_exec_function(PLpgSQL_function *func, FunctionCallInfo fcinfo,
 	int			rc;
+	bool orig_gp_enable_gpperfmon = gp_enable_gpperfmon;
+
+	if (log_min_messages > DEBUG5)
```
Why DEBUG5 in opposite to other places with DEBUG4?

dnskvlnk replied:
Fixed

Commit:
Fix gp_enable_gpperfmon in case of error

Commit:
Fix gp_enable_gpperfmon logging level

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/52512

Stolb27 commented on src/pl/plpgsql/src/pl_exec.c:
```diff
@@ -240,6 +240,7 @@ static char *format_expr_params(PLpgSQL_execstate *estate,
 static char *format_preparedparamsdata(PLpgSQL_execstate *estate,
 						  const PreparedParamsData *ppd);
 
+bool orig_gp_enable_gpperfmon;
```
Why is this variable in global scope? Do we need to change it outside this module?

dnskvlnk replied:
It's in global scope so that we can use it inside `plpgsql_exec_error_callback`

Stolb27 commented on src/pl/plpgsql/src/pl_exec.c:
```diff
@@ -893,6 +902,7 @@ static void
 plpgsql_exec_error_callback(void *arg)
 {
 	PLpgSQL_execstate *estate = (PLpgSQL_execstate *) arg;
+	gp_enable_gpperfmon = orig_gp_enable_gpperfmon;
```
Are you sure that this callback will be called without exceptions?
Moreover, it seems that you reinvent ~~the wheel~~ logic from `_SPI_pquery`

dnskvlnk replied:
Yes, the idea was to reproduce the logic from other parts of gpdb where we save `gp_enable_gpperfmon` to restore it later, but the main difference is that `PG_TRY` is used there and `plpgsql_exec_error_callback` is used in plpgsql_exec_function. Accodring to `plpgsql_exec_error_callback`'s comment, it's purpose is the same is `PG_TRY`'s, it should catch `ereport`s, so it seems to me rather safe to restore `gp_enable_gpperfmon` there

Stolb27 replied:
As we discussed, we are going to reuse sql function logic with pgtry/pgcatch

dnskvlnk replied:
Removed `gp_enable_gpperfmon` assignment from `plpgsql_exec_error_callback` and moved to `plpgsql_call_handler` where various plpgsql functions are called

Stolb27 replied:
Logic was moved to _SPI_execute_plan

Commit:
Rework plpgsql handling

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/52703

Stolb27 commented on src/pl/plpgsql/src/pl_handler.c:
```diff
@@ -218,6 +218,12 @@ plpgsql_call_handler(PG_FUNCTION_ARGS)
+	if (log_min_messages > DEBUG4)
+	{
+		gp_enable_gpperfmon = false;
+	}
```
Move this logic to _SPI_execute_plan as was  discussed yesterday

Commit:
Move gp_enable_gpperfmon handling to spi.c

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/52726

Stolb27 commented on src/backend/executor/spi.c:
```diff
@@ -2606,91 +2622,78 @@ _SPI_pquery(QueryDesc *queryDesc, bool fire_triggers, int64 tcount)
+	if (log_min_messages > DEBUG4)
+	{
+		gp_enable_gpperfmon = false;
+	}
```
Do we really need this block here? It would be nice to store this comment one frame above.

dnskvlnk replied:
Moved to `_SPI_execute_plan` inside PG_TRY block

Stolb27 commented on src/backend/executor/spi.c:
```diff
@@ -2093,6 +2093,12 @@ _SPI_execute_plan(SPIPlanPtr plan, ParamListInfo paramLI,
+	if (log_min_messages > DEBUG4)
+	{
+		gp_enable_gpperfmon = false;
+	}
```
This block should be inside PG_TRY as far as if an error happens before PG_TRY, the GUC value isn't restored back (see function.c for example).

dnskvlnk replied:
Removed this block and moved it inside PG_TRY block with its comment

Stolb27 commented on src/backend/executor/spi.c:
```diff
@@ -2377,6 +2384,15 @@ _SPI_execute_plan(SPIPlanPtr plan, ParamListInfo paramLI,
+PG_CATCH();
+{
+	gp_enable_gpperfmon = orig_gp_enable_gpperfmon;
+	PG_RE_THROW();
```
As far as other changes are requested, align this code with outer part too.

dnskvlnk replied:
Fixed

Stolb27 reviewed (commented):
> [gpdb incorrectly logs statements inslide plpgsql functions because](https://github.com/arenadata/gpdb/pull/591/commits/8a4113918de52748c5729818586322886a851541)
only executor queries were handled by PG_TRY, so gp_enable_gpperfmon
was set and reset correctly for such type of queries, but not during
utility queries which lead to incorrect gp_enable_gpperfmon handling.
This patch covers utility queries with PG_TRY as well, setting
gp_enable_gpperfmon to correct value for utility queries. Aslo,
gp_enable_gpperfmon was immideately reset in fmgr_sql to its original
value which lead incorrect logging of child queries. This patch also
removes extra assignment

1. gpdb incorrectly logs statements inslide plpgsql functions => gpperfmon logs some statements inslide sql/plpgsql functions despite log level
2. We need to pay attention to the logic about muting gpperfmon metrics for queries inside functions first. It was done by switching GUC value depending on the log level.
3. Then gp_enable_gpperfmon
was set and reset correctly for such type of queries, but not during
utility queries => that may contain nested executor queries, that will be unexpectedly logged by gpperfmon.
4. This patch move GUC manipulations to cover utility statements too.
5. **Aslo** (typo),
gp_enable_gpperfmon was **immideately** (typo) reset in fmgr_sql to its original
value which lead ~~incorrect~~ to unexpected logging of child queries.

Commit:
Move gp_enable_gpperfmon check inside PG_TRY block

Commit:
Fix formatting for PG_CATCH block

Commit:
Fix plpgsql/sql nested queries logging

gpdb changes gp_enable_gpperfmon guc value depending on log level to
control which statements to log during query execution. In a general
case we don't want do log nested statements of plpgsql function, but
gpperfmon logs some despite of log level which leads to unwanted and
incorrect records in queries_history where such records miss query
text and submit time. This happened because only executor queries were
properly covered to have correct gp_enable_gpperfmon value, when
utility queries were executed with wrong value of this guc. This patch
covers utility queries as well, setting gp_enable_gpperfmon to correct
value for them.

Aslo, gp_enable_gpperfmon was immideately reset in fmgr_sql to its
original value which lead incorrect logging of child queries. This
patch removes extra assignment.

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-3914-1

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/52870

RekGRpth commented on src/backend/executor/spi.c:
```diff
@@ -2278,6 +2279,17 @@ _SPI_execute_plan(SPIPlanPtr plan, ParamListInfo paramLI,
 
 			dest = CreateDestReceiver(canSetTag ? DestSPI : DestNone);
 
+PG_TRY();
```
Why did you keep the indentation when you added PG_TRY, but when you removed PG_TRY you shifted the indentation?

dnskvlnk replied:
I didn't change indentation in the place where I added PG_TRY to keep diff at minimum, that's the way they use in gp frequently. Also, I changed indentation in the place where I removed PG_TRY because it would look weird if I'd keep it as it was

RekGRpth commented on src/backend/executor/spi.c:
```diff
@@ -2606,91 +2625,69 @@ _SPI_pquery(QueryDesc *queryDesc, bool fire_triggers, int64 tcount)
 		eflags = EXEC_FLAG_SKIP_TRIGGERS;
 
-	PG_TRY();
-	{
```
How about saving this line? This will ensure that the indents below do not change, and the declaration of variables will be at the beginning of the block.

dnskvlnk replied:
Good suggestion, thanks. Already applied

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-3914-1

Commit:
Reduce diff by bringing back braces around former PG_TRY block

Commit:
Fix formatting

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/54075

Commit:
Fix plpgsql/sql nested queries logging

gpdb changes gp_enable_gpperfmon guc value depending on log level
to control which statements to log during query execution. We don't
want to log nested statements of plpgsql function if log level is
other than DEBUG5, but gpperfmon logs some despite of log level.
This happened because only executor queries were properly covered
to have correct gp_enable_gpperfmon value, when utility queries
were executed with wrong value of this guc.

This patch covers utility queries as well, setting gp_enable_gpperfmon
to correct value for them. Though gpperfmon doesn't log utility
queries, gp_enable_gpperfmon can be set for nested queries.

Also, gp_enable_gpperfmon was immediately reset in fmgr_sql to its
original value which lead incorrect logging of child queries. This
patch removes extra assignment.

RekGRpth commented on src/backend/executor/spi.c:
```diff
@@ -2093,6 +2093,7 @@ _SPI_execute_plan(SPIPlanPtr plan, ParamListInfo paramLI,
 	ErrorContextCallback spierrcontext;
 	CachedPlan *cplan = NULL;
 	ListCell   *lc1;
+	bool orig_gp_enable_gpperfmon = gp_enable_gpperfmon;
```
I think you should store gp_enable_gpperfmon not in this place, but below, before PG_TRY, as done in the fmgr_sql function.

dnskvlnk replied:
Moved

Commit:
Move gp_enable_gpperfmon declaration

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/54135
